synapse/lib/remotelayer.py

```diff
@@ -10,64 +11,129 @@
 
 logger = logging.getLogger(__name__)
 
-PASSTHROUGHFUNCS = (
-    'commit', 'getBuidProps', 'getOffset', 'setOffset', 'stat', 'initdb', 'stor', 'splicelistAppend',
-)
-
 class RemoteLayer(s_layer.Layer):
     '''
     A layer retrieved over telepath.
+
+    Expected behaviors during remote proxy connection failures:
+
+        * a reconnect loop begins which will run until success or isfini
+        * calls that were already sent or are iterating responses raise an error
+        * new calls will pend for up to "readywait" seconds and wake on ready or timeout.
+
     '''
     confdefs = (  # type: ignore
-        ('remote:telepath', {'type': 'str', 'doc': 'Path to remote layer'}),
+        ('url', {'type': 'str', 'doc': 'Path to remote layer'}),
+        ('readywait', {'type': 'int', 'defval': 30, 'doc': 'Max time to wait for layer ready.'}),
     )
 
-    async def __anit__(self, dirn, readonly=False, teleurl=None):
+    async def __anit__(self, core, node):
 
-        await s_layer.Layer.__anit__(self, dirn, readonly=readonly)
+        await s_layer.Layer.__anit__(self, core, node)
 
         # a remote layer may never be revd
+        self.proxy = None
         self.canrev = False
 
-        if teleurl is None:
-            teleurl = self.conf.get('remote:telepath')
+        self.ready = asyncio.Event()
+        await self._fireTeleTask()
+        self.onfini(self._onRemoteLayerFini)
+
+    async def _onRemoteLayerFini(self):
+        if self.proxy is not None:
+            await self.proxy.fini()
+
+    async def _fireTeleTask(self):
+
+        if self.isfini:
+            return
+
+        self.ready.clear()
+        self.schedCoro(self._teleConnLoop())
+
+    async def _teleConnLoop(self):
 
-        self.teleurl = teleurl
+        while not self.isfini:
 
-        #self.path = self.conf.get('remote:telepath')
-        if self.teleurl is None:
-            raise s_exc.BadConfValu(mesg='remote:telepath must be url for remote layer')
+            try:
 
-        self.remote = await s_telepath.openurl(self.teleurl)
-        self.onfini(self.remote.fini)
+                turl = self.conf.get('url')
+                self.proxy = await s_telepath.openurl(turl)
+                self.proxy.onfini(self._fireTeleTask)
+                self.ready.set()
+                logger.warning(f'connected to remote layer: {turl}')
+                return
 
-        for funcname in PASSTHROUGHFUNCS:
-            setattr(self, funcname, getattr(self.remote, funcname))
+            except asyncio.CancelledError:
+                raise
+
+            except Exception as e:
```

Comment:
In other connection loops we've got some variable backoff timing in place (for things ilke OSError vs. generic errors) in order to make us wait a little bit longer in the event that a service goes down, for a quick restart, as an example.  We may want to consider this so we don't end up in a log spamming situation.

Reply:
Most of those back-off situations start with a much lower time loop.  In this case, trying to reconnect once a second is reasonably slow ( and we'd want it to come back online ASAP when it comes back ) so i think this is probably best.